design/metadata-handling.md

```diff
@@ -0,0 +1,456 @@
+- **index*Hex** : All the `index` functions can be suffixed with `Hex` to
+  get the same value in hexadecimal format.
+- **bareMetalHostMACByName**: takes a string as parameter and returns the MAC
+  address of the nic with the name matching the parameter. This function
+  operates over the list of NICs in the `status.hardwareDetails.NIC` field of
+  the BareMetalHost.
+
+
+The output of the controller would be a Metal3Data object,one per node linking to the
+Metal3DataTemplate object and the associated secrets
+
+The Metal3Data object would be:
+
+```yaml
+apiVersion: infrastructure.cluster.x-k8s.io/v1alpha4
+kind: Metal3Data
+metadata:
+  name: nodepool-1-0
+  namespace: default
+  ownerReferences:
+  - apiVersion: infrastructure.cluster.x-k8s.io/v1alpha4
+    controller: true
+    kind: Metal3DataTemplate
+    name: nodepool-1
+spec:
+  index: 0
+  metaData:
+    name: machine-1-metadata
+    namespace: default
+  networkData:
+    name: machine-1-metadata
+    namespace: default
+  metal3Machine:
+    name: machine-1
+    namespace: default
+```
+The Metal3Data will contain the index of this node, and links to the secrets
+generated and to the Metal3Machine using this Metal3Data object.
+
+If the Metal3DataTemplate object is updated, the generated secrets will not be
+updated, to allow for reprovisioning of the nodes in the exact same state as
+they were initially provisioned. Hence, to do an update, it is necessary to do
+a rolling upgrade of all nodes.
+
+The reconciliation of the Metal3DataTemplate object will also be triggered by
+changes on Metal3Machines. In the case that a Metal3Machine gets modified, if
+the `dataTemplate` references a Metal3DataTemplate, that object will be reconciled.
+There will be two cases:
+
+- An already generated Metal3Data object exists with an ownerReference to this
+  Metal3Machine. In that case, the reconciler will verify that the required
+  secrets exist. If they do not, they will be created.
+- if no secret exists with an ownerReference to this Metal3Machine, then the
+  reconciler will create one and fill the respective field with the secret name.
+
+To create a Metal3Data object, the controller will select an index for that
+Metal3Machine. The selection happens by selecting the lowest available index.
+The controller will list all existing Metal3Data object linked to this
+Metal3DataTemplate and create a list of unavailable indexes. It will fill it by
+extracting the index from the Metal3Data names.
+
+Once the next available index is found, it will create the Metal3Data object.
+The name would be a concatenation of the Metal3DataTemplate name and index.
+Upon conflict, it will fetch again the list to consider the new list of 
+Metal3Data. Upon success, it will render the content values, and create
+the secrets containing the rendered data. The controller will generate the
+content based on the `metaData` or `networkData` field of the Metal3DataTemplate
+Specs.
```

Comment:
If the templates are only rendered at the point when the host is associated with the machine, then the secrets could also be named based on the machine name to ensure no conflict.

If we need an "index" for a machine, then could the thing creating the machine add an index to its metadata somehow? Normally I would think that is the MachineSet controller, but I don't know exactly how that works in the latest version of the APIs so I'm not sure if we have a place to plug in when the Machine is created. Doing it from the MachineSet avoid any conflicts, because only 1 goroutine should be reconciling the MachineSet and creating Machines for it.

Reply:
yes, the secrets name will contain the metal3machine name.

About the index, I am not sure that doing it in the machineset controller would be optimal, considering that it is bringing a requirement that is specific to Metal3 into CAPI impacting all providers. Considering this proposal, there also should be only one controller creating the indexes, the controller that reconciles the Metal3Data. No other controller should be doing this.